File: clonest1ck-haskell+python/day13.py
def collide(t, r):
    if(r == 0):
        return False
    if(t % (2 * (r - 1)) == 0):
        return True
    return False

def severity(firewall, time, break_on_sev):
    severity = 0
    level = 0
    while(level < len(firewall)):
        if(collide(time, firewall[level])):
            severity += level * firewall[level]
            if(break_on_sev):
                return 1
        level += 1
        time += 1

    return severity

f = open("in13.txt", 'r')
f = [x.split("\n")[0].split(": ") for x in f.readlines()]

firewall = []

for y in range(int(f[-1][0]) + 1):
    for x in f:
        if(int(x[0]) == y):
            firewall.append(int(x[1]))
    if(len(firewall) != y + 1):
        firewall.append(0)

print "Task 1: " + str(severity(firewall, 0, False))

delay = 0
while(severity(firewall, delay, True) != 0):
    delay += 1

print "Task 2: " + str(delay)

File: digidis-go/day13/day13.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type layer struct {
	Depth   int
	Range   int
	Scanner int
	dir     int
}

func main() {

	delay := 0

	f, _ := os.Open("input.txt")
	defer f.Close()
	scanner := bufio.NewScanner(f)
	layers := make([]*layer, 100)
	maxDepth := 0

	for scanner.Scan() {
		p := strings.Split(scanner.Text(), ": ")
		d, _ := strconv.Atoi(p[0])
		r, _ := strconv.Atoi(p[1])
		layers[d] = &layer{Depth: d, Range: r, Scanner: 0, dir: 0}
		if d > maxDepth {
			maxDepth = d
		}
	}
	f.Close()

	for {

		// set up scanner positions
		for _, l := range layers {
			if l != nil {
				s := delay % ((l.Range - 1) * 2)
				switch {
				case s < l.Range-1:
					l.Scanner = s
					l.dir = 0
				case s >= l.Range-1:
					l.Scanner = 2*(l.Range-1) - s
					l.dir = 1
				}
			}
		}

		pos := -1
		caught := 0
		severity := 0

		for pos < maxDepth {

			pos++
			l := layers[pos]
			if l != nil && l.Scanner == 0 {
				caught++
				if delay > 0 {
					break // abort early
				}
				severity += l.Depth * l.Range
			}

			//advance scanners
			for _, l := range layers {
				if l == nil {
					continue
				}
				if l.dir == 0 {
					l.Scanner++
					if l.Scanner == l.Range-1 {
						l.dir = 1
					}
				} else {
					l.Scanner--
					if l.Scanner == 0 {
						l.dir = 0
					}
				}
			}
		}

		if delay == 0 {
			fmt.Printf("Times caught %v\n", caught)
			fmt.Printf("Severity %v\n", severity)
		}
		if caught == 0 {
			fmt.Printf("Escaped after delaying %v\n", delay)
			break
		}
		if delay%100000 == 0 {
			fmt.Printf("%v\n", delay)
		}
		delay++

	}

}
